lib/dynflow/clock.rb

```diff
@@ -84,43 +70,31 @@ def add_timer(timer)
 
     def run_ready_timers
       while first_timer && first_timer.when <= Time.now
-        first_timer.apply
-        @timers.delete(first_timer)
+        begin
+          first_timer.apply
+        rescue => e
+          @logger && @logger.error("Failed to apply clock event #{first_timer}, exception: #{e}")
+        end
+        @timers.pop
       end
     end
 
     def first_timer
-      @timers.first
+      @timers.top
     end
 
     def wakeup
-      while @sleep_barrier.synchronize { Pill === @sleeping_pill }
-        Thread.pass
-      end
-      @sleep_barrier.synchronize do
-        @sleeper.wakeup if Took === @sleeping_pill
-      end
+      @timer.stop if @timer
     end
 
     def sleep_to(timer)
       return unless timer
       sec = [timer.when - Time.now, 0.0].max
-      @sleep_barrier.synchronize do
-        @sleeping_pill = Pill[sec]
-        @sleeper.wakeup
-      end
-    end
-
-    def sleeping
-      @sleep_barrier.synchronize do
-        loop do
-          @sleeping_pill = None
-          @sleep_barrier.sleep
-          pill           = @sleeping_pill
-          @sleeping_pill = Took
-          @sleep_barrier.sleep pill.value
-          reference.tell(:tick)
-        end
+      @timer = Concurrent::Channel.timer(sec)
```

Comment:
Any reason for using Channel vs. [ScheduledTask](http://ruby-concurrency.github.io/concurrent-ruby/master/Concurrent/ScheduledTask.html)?

Reply:
We may need to wake the timer earlier than it was scheduled to, `Channel.timer` provides nicer interface for that